pkg/controller/admissionchecks/multikueue/remoteclient.go

```diff
@@ -0,0 +1,179 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multikueue
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+)
+
+type clientWithWatchBuilder func(config []byte, options client.Options) (client.WithWatch, error)
+
+type remoteController struct {
+	localClient   client.Client
+	watchCtx      context.Context
+	watchCancel   context.CancelFunc
+	remoteClients map[string]*remoteClient
+	wlUpdateCh    chan<- event.GenericEvent
+
+	// For unit testing only. There is now need of creating fully functional remote clients in the unit tests
+	// and creating valid kubeconfig content is not trivial.
+	// The full client creation and usage is validated in the integration and e2e tests.
+	builderOverride clientWithWatchBuilder
+}
+
+func newRemoteController(watchCtx context.Context, localClient client.Client, wlUpdateCh chan<- event.GenericEvent) *remoteController {
+	watchCtx, watchCancel := context.WithCancel(watchCtx)
+	ret := &remoteController{
+		localClient: localClient,
+		watchCtx:    watchCtx,
+		watchCancel: watchCancel,
+		wlUpdateCh:  wlUpdateCh,
+	}
+	return ret
+}
+
+func (rc *remoteController) UpdateConfig(kubeConfigs map[string][]byte) error {
+	if rc.remoteClients == nil {
+		rc.remoteClients = make(map[string]*remoteClient, len(kubeConfigs))
+	}
+
+	for clusterName, c := range rc.remoteClients {
+		if kubeconfig, found := kubeConfigs[clusterName]; found {
+			if err := c.setConfig(kubeconfig); err != nil {
+				delete(rc.remoteClients, clusterName)
+				return fmt.Errorf("cluster %q: %w", clusterName, err)
+			}
+		} else {
+			c.watchCancel()
+			delete(rc.remoteClients, clusterName)
+		}
+	}
+	// create the missing ones
+	for clusterName, kubeconfig := range kubeConfigs {
+		if _, found := rc.remoteClients[clusterName]; !found {
+			c := newRemoteClient(rc.watchCtx, rc.localClient, rc.wlUpdateCh)
+
+			c.builderOverride = rc.builderOverride
+
+			if err := c.setConfig(kubeconfig); err != nil {
+				return fmt.Errorf("cluster %q: %w", clusterName, err)
+			}
+			rc.remoteClients[clusterName] = c
+		}
+	}
+
+	return nil
+}
+
+func (cc *remoteController) IsActive() bool {
+	return cc != nil && len(cc.remoteClients) > 0
+}
+
+type remoteClient struct {
+	localClient  client.Client
+	client       client.WithWatch
+	wlUpdateCh   chan<- event.GenericEvent
+	rootWatchCtx context.Context
+	watchCancel  context.CancelFunc
+	watchItf     watch.Interface
```

Comment:
What does `witf` stand for? Please name it more suggestively.

Reply:
watchItf?

Reply:
yes

Reply:
Oh, I suppose you meant `watchInterface`, I think calling it just `watch` or `watcher` is enough, but this is a non-blocking nit to me, wanted to know, because got confused

Reply:
reworked